src/model.rs

```diff
@@ -15,23 +15,42 @@ use serde::{Deserialize, Serialize};
     Ord,
 )]
 #[serde(rename_all = "snake_case")]
+// API reports; unknown variant `blabla`, expected one of
+// * `claude-3-5-sonnet-latest`,
+// * `claude-3-5-sonnet-20240620`,
+// * `claude-3-sonnet-20241022`,
+// * `claude-3-opus-latest`,
+// * `claude-3-opus-20240229`,
+// * `claude-3-sonnet-20240229`,
+// * `claude-3-5-haiku-latest`,
+// * `claude-3-5-haiku-20241022`,
+// * `claude-3-haiku-20240307`,
+// * `claude-3-haiku-latest`
+//
+// But docs say that `claude-3-5-sonnet-20241022` is a valid model, and the API
+// does accept it. This appears to be a bug in the API. - mdegans
+// https://docs.anthropic.com/en/docs/about-claude/models
```

Comment:
> I thought this was the case too but it turns out the error is not from the API but from the parsing the response into Model.

It *seems to be* fixed now. In cases where the API returns `claude-3-5-sonnet-20241022`, it does get deserialized to `Sonnet35_20241022`.

I think I may be missing something. Check out my dev branch and see if you can repro. Anything returned from the API should deserialize or there's for sure a bug. If you can break it, feel free to write a test for it so it can't break again.

To run with your api key, put it on `api.key` in the crate root and run `cargo test -- --ignored` (and delete it afterward). Or you can submit a PR against the dev branch and I'll run it with mine.

> It would probably be good to also handle parsing result more directly, so to make it easier to distinguish between the API result and the parse error.

There is a `client::Error` like this intended for that:

```rust
/// [`Client`] error type.
#[derive(Debug, thiserror::Error)]
pub enum Error {
    /// HTTP error.
    #[error("HTTP error: {0}")]
    HTTP(#[from] reqwest::Error),
    /// Data could not be parsed.
    #[error("Parse error: {0}")]
    Parse(#[from] serde_json::Error),
    /// Anthropic error.
    #[error("Anthropic error: {0}")]
    Anthropic(#[from] AnthropicError),
    /// Unexpected response from the API. These should never happen unless the
    /// server is misbehaving (for example, returning a stream when a message is
    /// expected).
    #[error("Unexpected response: {message}")]
    #[allow(missing_docs)]
    UnexpectedResponse { message: &'static str },
}
```

But when using `reqwest`'s `json()` to deserialize, they wrap it in their _own_ error message so the `Error::Parse` variant is never used. It ends up in a `Error::HTTP` error which it's not really. That is indeed a bug.

What I'll do is deserialize like you were doing before I reverted the change with `serde_json::from_str`, and then the `Error` should work as intended. I wasn't sure of your reasoning then, but it makes sense now.

As far as logging code goes, I *think* i didn't add it because `reqwest` logs them at the debug level, but I could be misremembering. I don't want to double-log if that's the case. I don't have time right this second to fix all this but will get to it on Friday.

Reply:
> If the Parse Error is not properly being used, maybe we should file something with reqwest then.

Nah. It's fine. I just had assumed that `.json()` would fail with a `serde_json` error but `reqwest`, like this crate, wraps it in one of their errors. `.json()` gets the bod, does a `serde_json::from_slice` and then a `.map_err` to a `reqwest` error type. But I overlooked that last part. It's now fixed. Failure to parse the json will give a `Error::Parse` now with the `serde_json` error inside. That was the intended behavior.

> I couldn't get logging for reqwest to work even with trace but I'm all for limiting logging if it's duplicative.

I am not 100% sure either way. I have to investigate the issue more. The `log` feature was mostly for my print debugging while developing the crate and so logging might be inconsistent (too much in some areas, not at all in others). I'll look into this before `0.6`.

> Thanks again for handling this so quickly.

YW! Hope you enjoy the crate.